src/libraries/System.Reflection.Metadata/tests/Metadata/TypeNameTests.cs

```diff
@@ -87,6 +96,42 @@ public void ParserIsNotEnforcingRuntimeSpecificRules(string input)
             }
         }
 
+        [Theory]
+        [InlineData("Type+InnerType", (string[])["Type", "InnerType"])]
+        [InlineData("Type+InnerType+InnermostType", (string[])["Type", "InnerType", "InnermostType"])]
+        [InlineData("NotNested\\+Name", (string[])["NotNested\\+Name"])]
+        [InlineData("NameEndingInBackSlash\\\\+NestedName", (string[])["NameEndingInBackSlash\\\\", "NestedName"])]
+        public void NestedName(string input, string[] expectedNames)
+        {
+            TypeName parsed = TypeName.Parse(input.AsSpan());
+            int i = expectedNames.Length - 1;
+            while (true)
+            {
+                Assert.Equal(expectedNames[i], parsed.Name);
+                // Caling FullName trims the _fullName value of the instance to just this type's full name.
+                // Test calling Name again to ensure it's still correct.
+                _ = parsed.FullName;
+                Assert.Equal(expectedNames[i], parsed.Name);
+                if (!parsed.IsNested)
+                {
+                    break;
+                }
+                parsed = parsed.DeclaringType;
+                i--;
+            }
```

Comment:
nit: it would be nice to ensure that the loop checks all the expected names

```suggestion
            }
            Assert.Equal(0, i);
```

Reply:
Done.